scripts/ssh-agent-helper

```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/bash
+# Helper script for starting the ssh agent if needed and doing an ssh-add.
+# This will let anyone smoothly run github/ssh related scripts without multiple password prompts.
+# This script is intended to be sourced.
+# Sourcing this script lets ssh-agent set the proper environment variables it needs to run properly.
+#
+# Expected usage:
+#
+# source ssh-agent-helper
+
+# SSH agent check: return code is 1 if there are no identities, 2 if cannot connect to agent.
+# Only start the agent on return code 2, otherwise we can just add our identity.
+# On return code 0 we don't have to do anything, the user already has this set up.
+ssh-add -L &> /dev/null
+rval=$?
+if [ "$rval" -eq 2 ]; then
+    echo "Starting ssh agent"
+    eval "$(ssh-agent -s)" &> /dev/null
+    ssh_agent_helper_cleanup() {
+        if [ -n "${SSH_AGENT_PID}" ]; then
+            echo "Cleaning up SSH agent"
+            eval "$(ssh-agent -k)" &> /dev/null
+        fi
+    }
+    trap ssh_agent_helper_cleanup EXIT
```

Comment:
This is some new bash-fu to me.

The idea here is that we start the ssh-agent, then kill it if `SSH_AGENT_PID` doesn't exist at the end of this script?  Doesn't the agent need to be running for keys to be applied?

What sets the `SSH_AGENT_PID` env var?  I don't have it in any of my sessions, with a forwarded agent at least.

Reply:
This `trap` line will call `ssh_agent_helper_cleanup` when the user's shell exits (the shell that sources this script). The intent of that function is to kill the ssh agent when the user logs out, because otherwise the agent process persists on the server indefinitely.

`SSH_AGENT_PID` is set by `eval "$(ssh-agent -s)` when you run it on the local host, it doesn't exist if you forward an agent. When you call `ssh-agent -s` it dumps a bunch of environment variable set commands to stdout. It's really weird but that's how it's coded. Then, you `eval` this output to get the environment variables that `ssh` needs to properly use the agent for your key.

If the `SSH_AGENT_PID` is no longer set, it's because someone already cleaned up the agent. In fact, without this environment variable set, the agent kill command doesn't know which agent to kill.

Some of this comment should be added to the script for clarity/documentation purposes.

Reply:
oh weird.  the man page for `ssh-agent` doesn't mention any env vars in the `-s` option documentation, and I'd never expect a command to just dump other commands to stdout 😕

Reply:
It's sort of mentioned in the man page but not in a very clear way. It doesn't expand much on what it's really doing.
```
There are two main ways to get an agent set up: The first is that the agent starts a new subcommand into which some environment variables are exported, eg ssh-agent xterm &.  The second is that the agent prints the needed shell commands (either sh(1) or csh(1) syntax can
be generated) which can be evaluated in the calling shell, eg eval `ssh-agent -s` for Bourne-type shells such as sh(1) or ksh(1) and eval `ssh-agent -c` for csh(1) and derivatives.
```